### PR TITLE
BUG: Dont include deleted rows from sas7bdat files (#15963)

### LINKED ISSUE #15963: IO/SAS (sas7bdat) deleted observations are not filtered out
#### Problem description

I filled a table, then deleted two observations (rows).
image
[datetime.sas7bdat](https://yadi.sk/d/h1En7L9M3GpDCV)
Pandas reads 5 rows with `read_sas`:
```
       Date1      Date2            DateTime                 DateTimeHi  \
0 1960-01-06 1960-01-04 1677-09-21 00:12:44 1677-09-21 00:12:43.145226   
1 1960-01-03 1960-01-05 2262-04-11 23:47:16 1960-01-01 00:00:00.000000   
2        NaT        NaT                 NaT                        NaT   
3 1960-01-06 1960-01-04 1677-09-21 00:12:44 2262-04-11 23:47:16.854774   
4 1960-01-01 1960-01-01                 NaT 1960-01-01 00:00:00.000000   

        Taiw  
0 1912-01-01  
1 1960-01-02  
2        NaT  
3 1912-01-01  
4 1960-01-01  
```

#### Expected Output

DataFrame with 3 rows instead of 5 rows (w/o index no.2 and no.4)

#### Output of ``pd.show_versions()``
last commit is cd24fa9 (ENH: add origin to to_datetime)
<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.3.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 58 Stepping 9, GenuineIntel
byteorder: little
LC_ALL: None
LANG: en
LOCALE: None.None

pandas: 0+unknown
pytest: 3.0.6
pip: 9.0.1
setuptools: 34.3.2
Cython: 0.25.2
numpy: 1.11.3
scipy: 0.19.0
xarray: 0.9.1
IPython: 5.3.0
sphinx: 1.5.2
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2016.10
blosc: 1.5.0
bottleneck: 1.2.0
tables: 3.3.0
numexpr: 2.6.2
feather: None
matplotlib: 2.0.0
openpyxl: None
xlrd: 1.0.0
xlwt: None
xlsxwriter: 0.9.5
lxml: 3.7.3
bs4: 4.5.3
html5lib: 1.0b10
sqlalchemy: 1.1.6
pymysql: None
psycopg2: None
jinja2: 2.9.5
s3fs: 0.0.9
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by troels**
Sas7bdat files may contain rows which are actually deleted.
 
After reverse engineering the file format a bit, I found that   
if the page_type has bit 7 set, there is a bitmap following
the normal row data with a bit set for a given row if it has been
deleted. Use that information to not include deleted rows in
the resulting dataframe.

This PR is built on top of #22628

I've added two test-cases one from the issue  and another constructed
example, which tests parsing on different page types and across multiple sas7bdat-pages. 

The constructed example is rather large, mainly to have a test case where the data 
flows across several pages.

- [X] closes #15963
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

### CODE REVIEW

**Comment by troels** [comment]
Hi @WillAyd  I made a separate PR #22628 for the two first commits. This one I though was large enough to have its own PR, but as there was a dependency on the first two commits I continued this PR from the branch #22628 had sprung from. 
If #22628 can be merged soon, this PR will suddenly get slightly smaller (but just a bit)

File: pandas/io/sas/sas.pyx [hunk 1]
```diff
@@ -309,10 +311,55 @@ cdef class Parser(object):
...
+        if self.parser._current_page_deleted_rows_bitmap_offset is None:
+            return -1
+
+        deleted_rows_bitmap_offset = \
+            self.parser._current_page_deleted_rows_bitmap_offset
+
+        page_type = self.current_page_type
+        subheader_pointers_length = \
+            self.subheader_pointer_length * self.current_page_subheaders_count
+
```
Comment (jreback) [inline pandas/io/sas/sas.pyx]: can you add some comments in here

File: pandas/io/sas/sas.pyx [hunk 2]
```diff
@@ -321,11 +368,29 @@ cdef class Parser(object):
         self.current_page_subheaders_count =\
             self.parser._current_page_subheaders_count
 
+        self.current_page_deleted_rows_bitmap_offset =\
+            self.calculate_deleted_rows_bitmap_offset()
+
+    cdef bint is_row_deleted(self, int row_number):
+        cdef:
```
Comment (jreback) [inline pandas/io/sas/sas.pyx]: doc-string & comments

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by WillAyd** [comment]
Can you address comments and merge master?

**Comment by jreback** [comment]
can you merge master and address comments

**Comment by WillAyd** [comment]
Closing as stale. Ping if you'd like to continue

**Comment by bezdomniy** [comment]
This would be pretty useful as the issue makes read_sas unusable for a lot of sas datasets.

**Pull request closed without merging**